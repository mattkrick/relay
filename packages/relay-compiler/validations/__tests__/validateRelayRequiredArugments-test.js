/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @format
 * @emails oncall+relay
 */

'use strict';

const GraphQLCompilerContext = require('../../core/GraphQLCompilerContext');
const RelayIRTransforms = require('../../core/RelayIRTransforms');

const validateRelayRequiredArugments = require('../validateRelayRequiredArugments');

const {transformASTSchema} = require('../../core/ASTConvert');
const {
  TestSchema,
  generateTestsFromFixtures,
  parseGraphQLText,
} = require('relay-test-utils-internal');
describe('validateRelayRequiredArugments-test', () => {
  const relaySchema = transformASTSchema(
    TestSchema,
    RelayIRTransforms.schemaExtensions,
  );

  generateTestsFromFixtures(
    `${__dirname}/fixtures/required-arguments`,
    text => {
      const {definitions, schema} = parseGraphQLText(relaySchema, text);
      const codegenContext = new GraphQLCompilerContext(TestSchema, schema)
        .addAll(definitions)
        .applyTransforms([
          ...RelayIRTransforms.commonTransforms,
          ...RelayIRTransforms.queryTransforms,
          ...RelayIRTransforms.printTransforms,
        ]);
      validateRelayRequiredArugments(codegenContext);
      return 'PASSED';
    },
  );
});
