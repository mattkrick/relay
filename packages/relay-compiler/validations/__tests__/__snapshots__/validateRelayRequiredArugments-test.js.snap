// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validateRelayRequiredArugments-test matches expected output: argument-on-field.graphql 1`] = `
~~~~~~~~~~ INPUT ~~~~~~~~~~
query TestQuery {
  node {
    hometown{
      nameWithArgs(capitalize: true)
    }
  }
}

~~~~~~~~~~ OUTPUT ~~~~~~~~~~
PASSED
`;

exports[`validateRelayRequiredArugments-test matches expected output: argument-on-linked-field.graphql 1`] = `
~~~~~~~~~~ INPUT ~~~~~~~~~~
fragment SomeComment on Comment {
  commentBody(supported: "Mark") {
    ... on PlainCommentBody {
      text {
        text
      }
    }
  }
}

query ViewerQuery {
  node {
    ...SomeComment
  }
}

~~~~~~~~~~ OUTPUT ~~~~~~~~~~
PASSED
`;

exports[`validateRelayRequiredArugments-test matches expected output: auto-filled-argument-on-defer.graphql 1`] = `
~~~~~~~~~~ INPUT ~~~~~~~~~~
query QueryWithFragmentWithDefer {
  me {
    id
    ...UserFragment
  }
}

fragment UserFragment on User {
  id
  emailAddresses
  ...DeferredFragment @defer
}

fragment DeferredFragment on User {
  id
  name
}

~~~~~~~~~~ OUTPUT ~~~~~~~~~~
PASSED
`;

exports[`validateRelayRequiredArugments-test matches expected output: auto-filled-argument-on-match.graphql 1`] = `
~~~~~~~~~~ INPUT ~~~~~~~~~~
fragment NameRendererFragment on User {
  id
  nameRendererForContext(context: HEADER) @match {
    ...PlainUserNameRenderer_name @module(name: "PlainUserNameRenderer.react")
    ...MarkdownUserNameRenderer_name
      @module(name: "MarkdownUserNameRenderer.react")
  }
}

fragment PlainUserNameRenderer_name on PlainUserNameRenderer {
  plaintext
  data {
    text
  }
}

fragment MarkdownUserNameRenderer_name on MarkdownUserNameRenderer {
  markdown
  data {
    markup
  }
}

~~~~~~~~~~ OUTPUT ~~~~~~~~~~
PASSED
`;

exports[`validateRelayRequiredArugments-test matches expected output: missing-argument-on-directive.invalid.graphql 1`] = `
~~~~~~~~~~ INPUT ~~~~~~~~~~
# expected-to-throw

query ViewerQuery {
  node {
    body @customDirective {
      text
    }
  }
}

~~~~~~~~~~ OUTPUT ~~~~~~~~~~
THROWN EXCEPTION:

Error: Encountered 1 error(s):
- Required argument 'level: Int!' is missing on 'customDirective' in 'ViewerQuery'.
  
  Source: GraphQL request (5:10)
  4:   node {
  5:     body @customDirective {
              ^
  6:       text
  
  Source: GraphQL request (3:1)
  2: 
  3: query ViewerQuery {
     ^
  4:   node {
  
`;

exports[`validateRelayRequiredArugments-test matches expected output: missing-argument-on-field.invalid.graphql 1`] = `
~~~~~~~~~~ INPUT ~~~~~~~~~~
# expected-to-throw
query TestQuery {
  node {
    hometown{
      nameWithArgs
    }
  }
}

~~~~~~~~~~ OUTPUT ~~~~~~~~~~
THROWN EXCEPTION:

Error: Encountered 1 error(s):
- Required argument 'capitalize: Boolean!' is missing on 'nameWithArgs' in 'TestQuery'.
  
  Source: GraphQL request (5:7)
  4:     hometown{
  5:       nameWithArgs
           ^
  6:     }
  
  Source: GraphQL request (2:1)
  1: # expected-to-throw
  2: query TestQuery {
     ^
  3:   node {
  
`;

exports[`validateRelayRequiredArugments-test matches expected output: missing-argument-on-linked-field.invalid.graphql 1`] = `
~~~~~~~~~~ INPUT ~~~~~~~~~~
# expected-to-throw
fragment SomeComment on Comment {
  commentBody {
    ... on PlainCommentBody {
      text {
        text
      }
    }
  }
}

query ViewerQuery {
  node {
    ...SomeComment
  }
}

~~~~~~~~~~ OUTPUT ~~~~~~~~~~
THROWN EXCEPTION:

Error: Encountered 1 error(s):
- Required argument 'supported: [String!]!' is missing on 'commentBody' in 'SomeComment'.
  
  Source: GraphQL request (3:3)
  2: fragment SomeComment on Comment {
  3:   commentBody {
       ^
  4:     ... on PlainCommentBody {
  
  Source: GraphQL request (2:1)
  1: # expected-to-throw
  2: fragment SomeComment on Comment {
     ^
  3:   commentBody {
  
`;
